Fix intrinsics treatment in Crossgen2

trylek opened the pull request:
This change fixes unintentional differences in the treatment of
intrinsics in Crossgen2 compared to Crossgen1 I found while
investigating failures in the regression test b426654. While the
change by itself doesn't fully fix the test, I believe it to be
generally useful towards removing the remaining codegen
differences between the compilers.

Thanks

Tomas

trylek committed:
Fix intrinsics treatment in Crossgen2

This change fixes unintentional differences in the treatment of
intrinsics in Crossgen2 compared to Crossgen1 I found while
investigating failures in the regression test b426654. While the
change by itself doesn't fully fix the test, I believe it to be
generally useful towards removing the remaining codegen
differences between the compilers.

Thanks

Tomas

jkotas commented:
We want to retire the old way of identifying intrinsics using hardcoded list and `CORINFO_FLG_INTRINSIC`.

I thought that we have done a pass to make sure all old-style intrinsics are marked with `[Intrinsic]`, but we are clearly missing some.

Could you please fix this by marking all old-style intrinsics (the ones with matching `CORINFO_INTRINSIC_XXX` constants) with `[Intrinsic]`?

trylek commented:
Sure, will do. Does this also mean we should ultimately deplete the contents of the method InitializeIntrinsicsHashtable in

https://github.com/dotnet/runtime/blob/cd622cd50c7104fc14a25fb199f8377472920f2e/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.Intrinsics.cs#L76

?

jkotas commented:
> we should ultimately deplete the contents of the method InitializeIntrinsicsHashtable 

Yes, this hashtable should be inside the JIT eventually.

MichalStrehovsky commented:
> I thought that we have done a pass to make sure all old-style intrinsics are marked

I made a pass over those that we generate special IL for but that was a different list (dotnet/coreclr#26209).

trylek commented:
OK, I'm about halfway through the refactoring suggested by @jkotas but I hit some difficulties and I have some extra questions:

(*) Changing the attribute on String.get_Chars, Type.GetTypeFromHandle or Object.GetType from InternalCall to Intrinsic crashes old crossgen when building SPC at the following line:

https://github.com/dotnet/runtime/blob/5bee16dae8bdc44731542566a5343aba900bbad2/src/coreclr/src/vm/methodtablebuilder.cpp#L4912

Should I change or remove the condition or keep both attributes on these methods?

(*) I cannot find where the three Array methods (Get, Set, Address) reside.

(*) Apparently there are many more internalcall methods than in the CorInfoImpl.Intrinsics set. According to the current code there will continue to be a difference between Crossgen1 and 2 w.r.t. their inlining. What is the right way to treat them? Should I keep the check for both attributes in Crossgen2 or add the Intrinsic attribute next to the InternalCall MethodImpl attribute for these methods?

Thanks a lot

Tomas

jkotas commented:
>  internalcall methods than in the CorInfoImpl.Intrinsics

internalcall methods that are not in the CorInfoImpl.Intrinsics set are opaque to the JIT. The JIT won't inline them nor make any other special assumptions about them.

trylek committed:
Address JanK's PR feedback

I have audited all entries in the intrinsics table and I added
the [Intrinsic] annotation to all applicable methods. I haven't
found <code>System.EETypePtr.EETypePtrOf</code> and
<code>System.Activator.DefaultConstructorOf</code> anywhere.
I have reverted the additional check for IsInternalCall that
I put in CorInfoImpl in the 1st commit.

Thanks

Tomas

trylek commented:
OK, I believe I have audited and fixed all the methods. I'm not completely sure about Interlocked.cs - for now I only marked the int variants as intrinsic, there's some mention in CorInfoImpl.Intrinsics about signature clashes among the various type-specific variants (long, float, double, IntPtr).

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.Intrinsics.cs:
```diff
@@ -179,7 +179,7 @@ private CorInfoIntrinsics getIntrinsicID(MethodDesc method, byte* pMustExpand)
                 *pMustExpand = 0;
 
-            Debug.Assert(method.IsIntrinsic);
+            Debug.Assert(method.IsIntrinsic || method.IsInternalCall);
```
This should be reverted.

trylek replied:
Reverted in 3rd commit, thanks for pointing that out!

jkotas commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -712,6 +712,11 @@ private uint getMethodAttribsInternal(MethodDesc method)
                 }
             }
 
+            if (method.HasCustomAttribute("System.Runtime.CompilerServices", "IntrinsicAttribute"))
```
Dtto

trylek replied:
Hmm, don't we need a check similar to what Crossgen1 has at

https://github.com/dotnet/runtime/blob/5bee16dae8bdc44731542566a5343aba900bbad2/src/coreclr/src/vm/methodtablebuilder.cpp#L5139

?

jkotas replied:
That's what `MethodDesc.IsIntrinsic` property does in crossgen2.

trylek replied:
OK, thanks for clarification, deleted in 4th commit. What about the CORINFO_FLG_JIT_INTRINSICS flag, should I also put it back? It seemed superfluous to me but I may be overlooking something.

jkotas replied:
Yes, you should put it back. Otherwise, the new-style intrinsics won't work.

trylek replied:
OK, put back in 5th commit, thanks for clarification.

trylek committed:
Revert assertion check for IsInternalCall

trylek committed:
Remove superfluous check for the intrinsic attribute

trylek committed:
Put back the CORINFO_FLG_JIT_INTRINSIC flag

trylek commented:
@jkotas - I believe I have addressed all your PR feedback and the tests seem to be more or less passing (there are a couple of remaining known issues in the Crossgen2 tests but none seem caused by my change); could you please take another look when you have a chance?

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/String.CoreCLR.cs:
```diff
@@ -26,6 +26,7 @@ public partial class String
         [System.Runtime.CompilerServices.IndexerName("Chars")]
         public extern char this[int index]
         {
+            [System.Runtime.CompilerServices.Intrinsic]
```
Nit: There is `using System.Runtime.CompilerServices` at the top of the file. This can be just `[Intrinsic]`

trylek replied:
Fixed in 6th commit.

jkotas reviewed (approved):
LGTM modulo one nit.

trylek committed:
Remove superfluous namespace name on Intrinsic per JanK's suggestion

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/String.CoreCLR.cs:
```diff
@@ -26,6 +26,7 @@ public partial class String
         [System.Runtime.CompilerServices.IndexerName("Chars")]
```
Nit: You can remove it here as well while you are on it

trylek replied:
Fixed in 7th commit.

trylek committed:
One more removal of superfluous namespace specification
